core/src/main/java/org/apache/iceberg/PositionDeletesTable.java

```diff
@@ -0,0 +1,252 @@
+                MetadataColumns.SPEC_ID_COLUMN_ID,
+                "spec_id",
+                Types.IntegerType.get(),
+                MetadataColumns.SPEC_ID_COLUMN_DOC),
+            Types.NestedField.required(
+                MetadataColumns.FILE_PATH_COLUMN_ID,
+                "delete_file_path",
+                Types.StringType.get(),
+                MetadataColumns.FILE_PATH_COLUMN_DOC));
+
+    if (partitionType.fields().size() > 0) {
+      return result;
+    } else {
+      // avoid returning an empty struct, which is not always supported.
+      // instead, drop the partition field
+      return TypeUtil.selectNot(result, Sets.newHashSet(MetadataColumns.PARTITION_COLUMN_ID));
+    }
+  }
+
+  public static class PositionDeletesBatchScan
+      extends SnapshotScan<BatchScan, ScanTask, ScanTaskGroup<ScanTask>> implements BatchScan {
+
+    protected PositionDeletesBatchScan(Table table, Schema schema) {
+      super(table, schema, new TableScanContext());
+    }
+
+    protected PositionDeletesBatchScan(Table table, Schema schema, TableScanContext context) {
+      super(table, schema, context);
+    }
+
+    @Override
+    protected PositionDeletesBatchScan newRefinedScan(
+        Table newTable, Schema newSchema, TableScanContext newContext) {
+      return new PositionDeletesBatchScan(newTable, newSchema, newContext);
+    }
+
+    @Override
+    public CloseableIterable<ScanTaskGroup<ScanTask>> planTasks() {
+      return TableScanUtil.planTaskGroups(
+          planFiles(), targetSplitSize(), splitLookback(), splitOpenFileCost());
+    }
+
+    @Override
+    protected List<String> scanColumns() {
+      return context().returnColumnStats() ? DELETE_SCAN_WITH_STATS_COLUMNS : DELETE_SCAN_COLUMNS;
+    }
+
+    @Override
+    protected CloseableIterable<ScanTask> doPlanFiles() {
+      String schemaString = SchemaParser.toJson(tableSchema());
+
+      // prepare transformed partition specs and caches
+      Map<Integer, PartitionSpec> transformedSpecs =
+          table().specs().values().stream()
+              .map(spec -> transformSpec(tableSchema(), spec))
+              .collect(Collectors.toMap(PartitionSpec::specId, spec -> spec));
+
+      LoadingCache<Integer, ResidualEvaluator> residualCache =
+          partitionCacheOf(
+              transformedSpecs,
+              spec ->
+                  ResidualEvaluator.of(
+                      spec,
+                      shouldIgnoreResiduals() ? Expressions.alwaysTrue() : filter(),
+                      isCaseSensitive()));
+
+      LoadingCache<Integer, String> specStringCache =
+          partitionCacheOf(transformedSpecs, PartitionSpecParser::toJson);
+
+      LoadingCache<Integer, ManifestEvaluator> evalCache =
+          partitionCacheOf(
+              transformedSpecs,
+              spec -> ManifestEvaluator.forRowFilter(filter(), spec, isCaseSensitive()));
+
+      // iterate through delete manifests
+      CloseableIterable<ManifestFile> deleteManifests =
```

Comment:
nit: This logic looks good but the Spotless formatting makes it a bit hard to read as lots of calls end up on separate lines, which makes them chunky. Maybe, adding empty lines between these calls and restructuring the code can help a bit.

```
List<ManifestFile> manifests = snapshot().deleteManifests(table().io());

CloseableIterable<ManifestFile> matchingManifests =
    CloseableIterable.filter(
        scanMetrics().skippedDeleteManifests(),
        CloseableIterable.withNoopClose(manifests),
        manifest -> evalCache.get(manifest.partitionSpecId()).eval(manifest));

matchingManifests =
    CloseableIterable.count(scanMetrics().scannedDeleteManifests(), matchingManifests);

Iterable<CloseableIterable<ScanTask>> tasks =
    CloseableIterable.transform(
        matchingManifests,
        manifest ->
            posDeletesScanTasks(
                manifest,
                tableSchemaString,
                transformedSpecs,
                residualCache,
                specStringCache));
```

This is completely up to you, @szehon-ho.